### ISSUE TITLE
FreeBSD 13.2 Build failures

### ISSUE THREAD

**Issue by LeeTibbert** [scala-native/scala-native#3540]
Bitrot. Compiled in January 2023. Passed almost all tests; failed 3, I think, network related tests (due to differing OS
specific device names).

FreeBSD 13.2 with current updates
Java 8
Scala Native 0.5.0-SNAPSHOT
Scala 3.3.1
LLVM-15
GC: commix, immix gives similar results. Have not tried Boehm
command: sbt> sandbox3/run

Something to peck away at whilst waiting for CI runs.  

This Issue may or may not be related to my recent series of `Discover.scala` Issues.

The `MAP_NORESERVE` probably will take the most study. 

```[info] Generating intermediate code (with debug metadata) (9259 ms)
[error] /usr/home/study_FreeBSD_VBox_2023-10-02T1544-0400/sandbox/.3/target/scala-3.3.1/native/native-code-nativelib_native0.5.0-SNAPSHOT_3-0.5.0-SNAPSHOT-0/scala-native/gc/shared/Safepoint.c:28:14: error: use of undeclared identifier 'MAP_NORESERVE'
[error]              MAP_NORESERVE | MAP_PRIVATE | MAP_ANONYMOUS, -1, 0);
[error]              ^
[error] 1 error generated.
[error] /usr/home/study_FreeBSD_VBox_2023-10-02T1544-0400/sandbox/.3/target/scala-3.3.1/native/native-code-nativelib_native0.5.0-SNAPSHOT_3-0.5.0-SNAPSHOT-0/scala-native/delimcc.c:60:9: warning: '__noinline' macro redefined [-Wmacro-redefined]
[error] #define __noinline __attribute__((noinline))
[error]         ^
[error] /usr/include/sys/cdefs.h:358:9: note: previous definition is here
[error] #define __noinline      __attribute__ ((__noinline__))
[error]         ^
[error] /usr/home/study_FreeBSD_VBox_2023-10-02T1544-0400/sandbox/.3/target/scala-3.3.1/native/native-code-nativelib_native0.5.0-SNAPSHOT_3-0.5.0-SNAPSHOT-0/scala-native/delimcc.c:62:26: error: use of undeclared identifier 'ASM_JMPBUF_SIZE'
[error] typedef void *lh_jmp_buf[ASM_JMPBUF_SIZE / sizeof(void *)];
[error]                          ^
[error] /usr/home/study_FreeBSD_VBox_2023-10-02T1544-0400/sandbox/.3/target/scala-3.3.1/native/native-code-nativelib_native0.5.0-SNAPSHOT_3-0.5.0-SNAPSHOT-0/scala-native/delimcc.c:303:43: error: use of undeclared identifier 'ASM_JMPBUF_SIZE'
[error]     memcpy(return_buf, continuation->buf, ASM_JMPBUF_SIZE);
[error]                                           ^
[error] /usr/home/study_FreeBSD_VBox_2023-10-02T1544-0400/sandbox/.3/target/scala-3.3.1/native/native-code-nativelib_native0.5.0-SNAPSHOT_3-0.5.0-SNAPSHOT-0/scala-native/delimcc.c:323:5: error: use of undeclared identifier 'JMPBUF_STACK_POINTER_OFFSET'
[error]     jmpbuf_fix(return_buf);
[error]     ^
[error] /usr/home/study_FreeBSD_VBox_2023-10-02T1544-0400/sandbox/.3/target/scala-3.3.1/native/native-code-nativelib_native0.5.0-SNAPSHOT_3-0.5.0-SNAPSHOT-0/scala-native/delimcc.c:322:38: note: expanded from macro 'jmpbuf_fix'
[error] #define jmpbuf_fix(buf) fix_addr(buf[JMPBUF_STACK_POINTER_OFFSET])
[error]                                      ^
[error] /usr/home/study_FreeBSD_VBox_2023-10-02T1544-0400/sandbox/.3/target/scala-3.3.1/native/native-code-nativelib_native0.5.0-SNAPSHOT_3-0.5.0-SNAPSHOT-0/scala-native/delimcc.c:323:5: error: use of undeclared identifier 'JMPBUF_STACK_POINTER_OFFSET'
[error] /usr/home/study_FreeBSD_VBox_2023-10-02T1544-0400/sandbox/.3/target/scala-3.3.1/native/native-code-nativelib_native0.5.0-SNAPSHOT_3-0.5.0-SNAPSHOT-0/scala-native/delimcc.c:322:38: note: expanded from macro 'jmpbuf_fix'
[error] #define jmpbuf_fix(buf) fix_addr(buf[JMPBUF_STACK_POINTER_OFFSET])
[error]                                      ^
[error] /usr/home/study_FreeBSD_VBox_2023-10-02T1544-0400/sandbox/.3/target/scala-3.3.1/native/native-code-nativelib_native0.5.0-SNAPSHOT_3-0.5.0-SNAPSHOT-0/scala-native/delimcc.c:333:9: error: use of undeclared identifier 'JMPBUF_STACK_POINTER_OFFSET'
[error]         jmpbuf_fix(nw->h->buf);
[error]         ^
[error] /usr/home/study_FreeBSD_VBox_2023-10-02T1544-0400/sandbox/.3/target/scala-3.3.1/native/native-code-nativelib_native0.5.0-SNAPSHOT_3-0.5.0-SNAPSHOT-0/scala-native/delimcc.c:322:38: note: expanded from macro 'jmpbuf_fix'
[error] #define jmpbuf_fix(buf) fix_addr(buf[JMPBUF_STACK_POINTER_OFFSET])
[error]                                      ^
[error] /usr/home/study_FreeBSD_VBox_2023-10-02T1544-0400/sandbox/.3/target/scala-3.3.1/native/native-code-nativelib_native0.5.0-SNAPSHOT_3-0.5.0-SNAPSHOT-0/scala-native/delimcc.c:333:9: error: use of undeclared identifier 'JMPBUF_STACK_POINTER_OFFSET'
[error] /usr/home/study_FreeBSD_VBox_2023-10-02T1544-0400/sandbox/.3/target/scala-3.3.1/native/native-code-nativelib_native0.5.0-SNAPSHOT_3-0.5.0-SNAPSHOT-0/scala-native/delimcc.c:322:38: note: expanded from macro 'jmpbuf_fix'
[error] #define jmpbuf_fix(buf) fix_addr(buf[JMPBUF_STACK_POINTER_OFFSET])
[error]                                      ^
[error] 1 warning and 6 errors generated.
[info] Compiling to native code (10661 ms)
[info] Total (44294 ms)
[warn] In the last 937 seconds, 12.866 (1.4%) were spent in GC. [Heap: 1.41GB free of 2.02GB, max 12.00GB] Consider increasing the JVM heap using `-Xmx` or try a different collector, e.g. `-XX:+UseG1GC`, for better performance.
[error] Failed to compile /usr/home/study_FreeBSD_VBox_2023-10-02T1544-0400/sandbox/.3/target/scala-3.3.1/native/native-code-nativelib_native0.5.0-SNAPSHOT_3-0.5.0-SNAPSHOT-0/scala-native/gc/shared/Safepoint.c
[error] (sandbox3 / Compile / nativeLink) Failed to compile /usr/home/study_FreeBSD_VBox_2023-10-02T1544-0400/sandbox/.3/target/scala-3.3.1/native/native-code-nativelib_native0.5.0-SNAPSHOT_3-0.5.0-SNAPSHOT-0/scala-native/gc/shared/Safepoint.c
[error] Total time: 941 s (15:41), completed Oct 5, 2023 9:34:32 PM
```

**Comment by LeeTibbert** [comment]
@WojciechMazur 

I am trying to verify my FreeBSD X64 build using the latest 0.5.0-SNAPSHOT

I am aware that am venturing into an OS not exercised on CI. That is my intent, 
if only to identify for end users where the broken parts are.  SN can not be all things
to all people. It is nice if a finite amount of effort makes the tent available to more people.

When I run `test-tools`, I get a series of lines I do not remember seeing on Linux 64 or Apple M1.
Off the top of your head, do I need to be concerned about these? "No idea!" is a reasonable answer.

(I can correct the typo s/ammount/amount, that is not my biggest problem at present). 
`test-tools` does not halt. If one ignores the three known failures I am in the process of fixing 
`test-runtime` passes.  I am running `test-scripted` now (that system is __slow__! I expect to
have results by the New Year).


```
[info] Test scala.scalanative.linker.MinimalRequiredSymbolsTest.default started

Ammount of found symbols in norm, config={debugMetadata=false, multithreading=f\
alse, targetTriple=x86_64-unknown-unknown}:
Expected at most: {types=710, members=3600, total=4310}
Found:            {types=705, members=3528, total=4233}
Diff:             {types=-5, members=-72, total=-77}
```
Repeated for multithread and perhaps a few others.

**Comment by LeeTibbert** [comment]
@natsukagami 

May I please enlist your help answering a few questions now and later reviewing a PR?

I am trying to restore Scala Native FreeBSD building and passing all tests. Some bitrot
has set in.  I have reached an area where you have far more experience than I.

In PR #3286 you added a number of files.  `delimcc.c` and `setjmp_amd64.S` were added.

From the text in `setjmp_amd64.S`, it looks like the PR code is derived from a prior source.
Merit to you for leaving the original text to make it easier for contributors like me to follow.
If that is true, do you remember its origin?  

It is hard to follow or backtrack the trail of clever people. Those files  seems not to be in the `libhandler` GitHub
repository you mentioned in the Issue. I did several Google searches for what I thought might be fruitful
terms, including the original author's name, but no luck.

I am interested in the history/reasoning of the line at the very top of the file:
```
#if defined(__x86_64__) && (defined(__linux__) || defined(__APPLE__))
```
From the text below it, it seems to me that block of code should also work on FreeBSD.
The FreeBSD setjmp definition has not changed in a decade or more, which means
the structure in the file should still work.

When I change that (note 1 below) to:
```
#if defined(__x86_64__) && \
        (defined(__linux__) || defined(__APPLE__) || defined(__FreeBSD__))
```
* `test-tools` succeeds (with possibly an unrelated set of warnings)
* `test-runtime` succeeds if one ignores the 3 know network related errors I am trying to fix
* `test-scripted` fails without providing any obvious information. 
    I think that failure is due to the limited amount of memory and disk space on my system.
    I would like to rule out any setjmp defects my hacking may have caused.

* I do not know if any continuations tests were actually run.  I suspect that they were not
   and that I need to enable something experimental (unit-tests-experimental or such).
   Kinda if you do not know the secret password, you are not eligible to come knocking on the door.

Thank you for your insights.

Note 1:  I made corresponding changes in `delimcc.c` and also an unrelated one in a GC file.
               So this is not PR ready.  I am aware that few other people are likely to have a FreeBSD
               system available.   I am in that situation myself with lacking a Windows development environment.

**Comment by natsukagami** [comment]
@LeeTibbert 

Thanks for trying to port it to FreeBSD! I have zero experience with it, so I cannot be sure about how applicable it is, however in theory it should be the same, should the compilers on FreeBSD generate the same C calling convention code and uses the same exception handling mechanism (for C++) as Linux does. 

> If that is true, do you remember its origin?

The setjmp/longjmp implementation indeed came from https://github.com/koka-lang/libhandler, however the rest of the functions are written myself. They should do fairly simple things and the functionality should be well documented inside `delimcc.c` where I define the ABI interface for them, here: https://github.com/scala-native/scala-native/blob/95742a05b942c4f7796b4fee43f411109ac97e97/nativelib/src/main/resources/scala-native/delimcc.c#L61-L76

**Comment by LeeTibbert** [comment]
@natsukagami Thank you for your helpful reply and its info.

I studied your original PR and found `Scala3ContinuationsTests`. 
I have been working on getting `test-runtime` to succeed. 

After that is doing the specific `ContinuationsTests`. The time
you put into writing them is going to save me __much__ aggravation.
Especially welcome since I have little knowledge in this area.

**Comment by LeeTibbert** [comment]
@devlaam  

Thank you for the encouragement. 
Always welcome, but especially welcome now. 

FreeBSD 14 is about to release and it will support 1024 (or more)? Cores.  Not that I have that many
cores available, but someone might and I would like Scala Native to be there, even if untested locally or CI.

If I might ask, are you running on X86_64 or arm64 or something else?   The reason that I ask is that
I am currently trying to get `test-all` to run cleanly on X86_64. Most of the unit-tests run, but I am getting
some strange hangs, arrgh!  Anyway, I hope to either fix or document those soon (days).

In January 2023, I had this all running and tested (with documented test failures) on arm64.  It may be
quite a while before I am able to spin up an arm64 machine.  If I knew there was a user in the wild who
might be able to use SN on FreeBSD arm64, I might move the arm64 priority up.

Thanks again.

**Comment by devlaam** [comment]
I am running this on X86_64, but the machine is not that young any more and I haven't decided what the new architecture will be. most likely  X86_64 though. But i think that you may also take into consideration what the major platforms for virtual computers will be in the future. People which are running their own hardware in a data warehouse are a very small minority i guess. 

Keep up the goog work! I understand it is a lonely journey right now, but I am convinced this will change once when the application space for Native increases.
